server/modules/nodeDef/manager/nodeDefManager.js

```diff
@@ -53,6 +55,49 @@ const updateNodeDefProps = async (user, surveyId, nodeDefUuid, props, propsAdvan
     return nodeDef
   })
 
+const deleteNodeDefValidationMessagesLabels = async (surveyId, deletedLanguages, client = db) => {
+  await client.tx(async tx => {
+    // 1. fetch all node defs (already published)
+    const nodeDefs = await NodeDefRepository.fetchNodeDefsBySurveyId(surveyId, false, true, tx)
+
+    for (const nodeDef of nodeDefs) {
+      // 2. for each node def, remove messages in deleted languages and update its expressions in validations
+      const validations = NodeDef.getValidations(nodeDef)
+      const expressions = NodeDefValidations.getExpressions(validations)
+
+      let nodeDefUpdateRequired = false
+      const expressionsUpdated = []
+
+      for (const expression of expressions) {
+        // 3. for each expression find messages in deleted languages (if any)
+        const messages = NodeDefExpression.getMessages(expression)
+
+        const deletedMessageLanguages = R.pipe(
+          R.keys,
+          R.intersection(deletedLanguages)
+        )(messages)
+
+        if (R.isEmpty(deletedMessageLanguages)) {
+          expressionsUpdated.push(expression)
+        } else {
+          // 4. remove messages in deleted languages
+          const messagesUpdated = R.omit(deletedMessageLanguages, messages)
+          expressionsUpdated.push(NodeDefExpression.assocMessages(messagesUpdated)(expression))
+          nodeDefUpdateRequired = true
+        }
+
+        if (nodeDefUpdateRequired) {
+          // 5. update node def in db (props already published)
+          const validationsUpdated = NodeDefValidations.assocExpressions(expressionsUpdated)(validations)
+          await NodeDefRepository.updateNodeDefPropsPublished(surveyId, NodeDef.getUuid(nodeDef), {}, {
```

Comment:
why did you introduce this ```NodeDefRepository .updateNodeDefPropsPublished```? Isn't the publishProps happening after this? don't you just need to update the expressions draft props here?

Reply:
in the other cases (survey labels/descriptions, nodeDef labels/descriptions, category items labels) it's easier to do it after survey props are published, so it's better to remove even the validation message labels together with them.
Also, NodeDefRepository.updateNodeDefProps updates even the date_modified, and it's not needed in this case.